### PR TITLE
[SPARK-44106][PYTHON][CONNECT] Add `__repr__` for `GroupedData`

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
Add `__repr__` for `GroupedData`

### Why are the changes needed?
 `GroupedData.__repr__` is missing

### Does this PR introduce _any_ user-facing change?
yes

1. On Scala side:
```
scala> val df = Seq(("414243", "4243")).toDF("e", "f")
df: org.apache.spark.sql.DataFrame = [e: string, f: string]

scala> df.groupBy("e")
res0: org.apache.spark.sql.RelationalGroupedDataset = RelationalGroupedDataset: [grouping expressions: [e: string], value: [e: string, f: string], type: GroupBy]

scala> df.groupBy(df.col("e"))
res1: org.apache.spark.sql.RelationalGroupedDataset = RelationalGroupedDataset: [grouping expressions: [e: string], value: [e: string, f: string], type: GroupBy]
```

2. On vanilla PySpark:

before this PR:
```
In [1]: df = spark.createDataFrame([("414243", "4243",)], ["e", "f"])

In [2]: df
Out[2]: DataFrame[e: string, f: string]

In [3]: df.groupBy("e")
Out[3]: <pyspark.sql.group.GroupedData at 0x10423a4c0>

In [4]: df.groupBy(df.e)
Out[4]: <pyspark.sql.group.GroupedData at 0x1041dd640>

```

after this PR:
```
In [1]: df = spark.createDataFrame([("414243", "4243",)], ["e", "f"])

In [2]: df
Out[2]: DataFrame[e: string, f: string]

In [3]: df.groupBy("e")
Out[3]: GroupedData[grouping expressions: [e], value: [e: string, f: string], type: GroupBy]

In [4]: df.groupBy(df.e)
Out[4]: GroupedData[grouping expressions: [e: string], value: [e: string, f: string], type: GroupBy]
```

3. On Spark Connect Python Client:
before this PR:
```
In [1]: df = spark.createDataFrame([("414243", "4243",)], ["e", "f"])

In [2]: df
Out[2]: DataFrame[e: string, f: string]

In [3]: df.groupBy("e")
Out[3]: <pyspark.sql.connect.group.GroupedData at 0x1046157c0>

In [4]: df.groupBy(df.e)
Out[4]: <pyspark.sql.connect.group.GroupedData at 0x11da5ceb0>
```

after this PR:
```
In [1]: df = spark.createDataFrame([("414243", "4243",)], ["e", "f"])

In [2]: df
Out[2]: DataFrame[e: string, f: string]

In [3]: df.groupBy("e")
Out[3]: GroupedData[grouping expressions: [e], value: [e: string, f: string], type: GroupBy]

In [4]: df.groupBy(df.e)
Out[4]: GroupedData[grouping expressions: [e], value: [e: string, f: string], type: GroupBy] // different from vanilla PySpark
```

Note that since the expressions in Python Client are not resolved, the string can be different from vanilla PySpark. 

### How was this patch tested?
added doctests

### CODE REVIEW

**Comment by zhengruifeng** [comment]
cc @HyukjinKwon

File: python/pyspark/sql/group.py [hunk 1]
```diff
@@ -70,6 +70,10 @@ def __init__(self, jgd: JavaObject, df: DataFrame):
         self._df = df
         self.session: SparkSession = df.sparkSession
 
+    def __repr__(self) -> str:
+        j_str = self._jgd.toString()
+        return "GroupedData" + j_str[26:]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/group.py:75]: qq what's this 26?

File: python/pyspark/sql/connect/group.py [hunk 2]
```diff
@@ -83,6 +83,26 @@ def __init__(
...
             self._pivot_values = pivot_values
 
+    def __repr__(self) -> str:
+        # since the expressions are not resolved here,
+        # the string represent can be different from vanilla PySpark.
+        grouping_str = ", ".join(str(e._expr) for e in self._grouping_cols)
+        grouping_str = f"grouping expressions: [{grouping_str}]"
+
+        value_str = str(self._df)
+        value_str = value_str[9:]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/group.py:93]: and this 9?

File: python/pyspark/sql/group.py (hunk 1 above)
Comment (zhengruifeng) [reply python/pyspark/sql/group.py:75]: it is used to truncate the string from JVM `RelationalGroupedDataset: [grouping expressions: [e: string], value: [e: string, f: string], type: GroupBy]`

File: python/pyspark/sql/connect/group.py (hunk 2 above)
Comment (zhengruifeng) [reply python/pyspark/sql/connect/group.py:93]: for this one, we can build it from the schema, let me update

File: python/pyspark/sql/group.py [hunk 3]
```diff
@@ -70,6 +70,9 @@ def __init__(self, jgd: JavaObject, df: DataFrame):
         self._df = df
         self.session: SparkSession = df.sparkSession
 
+    def __repr__(self) -> str:
+        return f"GroupedData{self._jgd.toString()[26:]}"
```
Comment (HyukjinKwon) [inline python/pyspark/sql/group.py:74]: I would put 26 to a separate variable, and add an assert if the truncated string starts with `[` ..

Comment (HyukjinKwon) [reply python/pyspark/sql/group.py:74]: If we'd be more conservative, we can invoke super if the string doesn't start with `[`.

Comment (zhengruifeng) [reply python/pyspark/sql/group.py:74]: let me check again

**Review by HyukjinKwon — APPROVED** [review]

**Comment by zhengruifeng** [comment]
thanks @HyukjinKwon for reviews!

merged into master

**Pull request closed without merging**